File: facebook/src/com/facebook/internal/PlatformServiceClient.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.internal;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.*;

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
abstract public class PlatformServiceClient implements ServiceConnection {
    private final Context context;
    private final Handler handler;
    private CompletedListener listener;
    private boolean running;
    private Messenger sender;
    private int requestMessage;
    private int replyMessage;
    private final String applicationId;
    private final int protocolVersion;

    public PlatformServiceClient(
            Context context,
            int requestMessage,
            int replyMessage,
            int protocolVersion,
            String applicationId) {
        Context applicationContext = context.getApplicationContext();

        this.context = (applicationContext != null) ? applicationContext : context;
        this.requestMessage = requestMessage;
        this.replyMessage = replyMessage;
        this.applicationId = applicationId;
        this.protocolVersion = protocolVersion;

        handler = new Handler() {
            @Override
            public void handleMessage(Message message) {
                PlatformServiceClient.this.handleMessage(message);
            }
        };
    }

    public void setCompletedListener(CompletedListener listener) {
        this.listener = listener;
    }

    protected Context getContext() {
        return context;
    }

    public boolean start() {
        if (running) {
            return false;
        }

        // Make sure that the service can handle the requested protocol version
        int availableVersion = NativeProtocol.getLatestAvailableProtocolVersionForService(
                protocolVersion);
        if (availableVersion == NativeProtocol.NO_PROTOCOL_AVAILABLE) {
            return false;
        }

        Intent intent = NativeProtocol.createPlatformServiceIntent(context);
        if (intent == null) {
            return false;
        } else {
            running = true;
            context.bindService(intent, this, Context.BIND_AUTO_CREATE);
            return true;
        }
    }

    public void cancel() {
        running = false;
    }

    public void onServiceConnected(ComponentName name, IBinder service) {
        sender = new Messenger(service);
        sendMessage();
    }

    public void onServiceDisconnected(ComponentName name) {
        sender = null;
        try {
            context.unbindService(this);
        } catch (IllegalArgumentException ex) {
            // Do nothing, the connection was already unbound
        }
        callback(null);
    }

    private void sendMessage() {
        Bundle data = new Bundle();
        data.putString(NativeProtocol.EXTRA_APPLICATION_ID, applicationId);

        populateRequestBundle(data);

        Message request = Message.obtain(null, requestMessage);
        request.arg1 = protocolVersion;
        request.setData(data);
        request.replyTo = new Messenger(handler);

        try {
            sender.send(request);
        } catch (RemoteException e) {
            callback(null);
        }
    }

    protected abstract void populateRequestBundle(Bundle data);

    protected void handleMessage(Message message) {
        if (message.what == replyMessage) {
            Bundle extras = message.getData();
            String errorType = extras.getString(NativeProtocol.STATUS_ERROR_TYPE);
            if (errorType != null) {
                callback(null);
            } else {
                callback(extras);
            }
            context.unbindService(this);
        }
    }

    private void callback(Bundle result) {
        if (!running) {
            return;
        }
        running = false;

        CompletedListener callback = listener;
        if (callback != null) {
            callback.completed(result);
        }
    }

    public interface CompletedListener {
        void completed(Bundle result);
    }
}

File: facebook-common/src/main/java/com/facebook/internal/PlatformServiceClient.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import android.content.ComponentName
import android.content.Context
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import com.facebook.internal.NativeProtocol.createPlatformServiceIntent
import com.facebook.internal.NativeProtocol.getLatestAvailableProtocolVersionForService
import java.lang.IllegalArgumentException

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
abstract class PlatformServiceClient(
    context: Context,
    requestMessage: Int,
    replyMessage: Int,
    protocolVersion: Int,
    applicationId: String,
    redirectURI: String,
    nonce: String?
) : ServiceConnection {
  protected val context: Context
  private val handler: Handler
  private var listener: CompletedListener? = null
  private var running = false
  private var sender: Messenger? = null
  private val requestMessage: Int
  private val replyMessage: Int
  private val applicationId: String
  private val redirectURI: String?
  private val protocolVersion: Int
  val nonce: String?

  fun setCompletedListener(listener: CompletedListener?) {
    this.listener = listener
  }

  fun start(): Boolean {
    synchronized(this) {
      if (running) {
        return false
      }

      // Make sure that the service can handle the requested protocol version
      val availableVersion = getLatestAvailableProtocolVersionForService(protocolVersion)
      if (availableVersion == NativeProtocol.NO_PROTOCOL_AVAILABLE) {
        return false
      }
      val intent = createPlatformServiceIntent(context)
      return if (intent == null) {
        false
      } else {
        running = true
        context.bindService(intent, this, Context.BIND_AUTO_CREATE)
        true
      }
    }
  }

  fun cancel() {
    running = false
  }

  override fun onServiceConnected(name: ComponentName, service: IBinder) {
    sender = Messenger(service)
    sendMessage()
  }

  override fun onServiceDisconnected(name: ComponentName) {
    sender = null
    try {
      context.unbindService(this)
    } catch (ex: IllegalArgumentException) {
      // Do nothing, the connection was already unbound
    }
    callback(null)
  }

  private fun sendMessage() {
    val data = Bundle()
    data.putString(NativeProtocol.EXTRA_APPLICATION_ID, applicationId)
    if (nonce != null) {
      data.putString(NativeProtocol.EXTRA_NONCE, nonce)
    }
    if (redirectURI != null) {
      data.putString(NativeProtocol.EXTRA_REDIRECT_URI, redirectURI)
    }
    populateRequestBundle(data)
    val request = Message.obtain(null, requestMessage)
    request.arg1 = protocolVersion
    request.data = data
    request.replyTo = Messenger(handler)
    try {
      sender?.send(request)
    } catch (e: RemoteException) {
      callback(null)
    }
  }

  protected abstract fun populateRequestBundle(data: Bundle)
  protected fun handleMessage(message: Message) {
    if (message.what == replyMessage) {
      val extras = message.data
      val errorType = extras.getString(NativeProtocol.STATUS_ERROR_TYPE)
      if (errorType != null) {
        callback(null)
      } else {
        callback(extras)
      }
      try {
        context.unbindService(this)
      } catch (ex: IllegalArgumentException) {
        // Do nothing, the connection was already unbound
      }
    }
  }

  private fun callback(result: Bundle?) {
    if (!running) {
      return
    }
    running = false
    val callback = listener
    callback?.completed(result)
  }

  fun interface CompletedListener {
    fun completed(result: Bundle?)
  }

  init {
    val applicationContext = context.applicationContext
    this.context = applicationContext ?: context
    this.requestMessage = requestMessage
    this.replyMessage = replyMessage
    this.applicationId = applicationId
    this.redirectURI = redirectURI
    this.protocolVersion = protocolVersion
    this.nonce = nonce
    handler =
        object : Handler() {
          override fun handleMessage(message: Message) {
            this@PlatformServiceClient.handleMessage(message)
          }
        }
  }
}
